Add support for %short title and %envname Resolves #748

MeanSquaredError opened the pull request:
This PR fixes the last test error reported in #748.
There was a report of a failing test
`FAIL! : LatexParsingTest::test_latexLexing(section command, multi-line optional)`
`'Token::generalArg == Token::UnknownTokenType' returned FALSE. (incorrect subtype at index 1)`

After some investigation it turned out that the error could be reproduced by enabling the checkbox for  yathesis.cwl in `Options / Configure TeXstudio / Completion / yathesis.cwl`

Initially I thought that it was a defective test but it turned out that the test is actually catching a real bug in the CWL code in yathesis.cwl
This file contains a number of cwl definitions like :
`\section[titre alt. pour TdM et entête%short title]{titre%title}#L2`
`\paragraph*{titre%title}#L5`
The problem with these definitions is that `%short title` and `%title` are not valid modifiers for the argument type so using them is ignored and the corresponding arguments get the type "general argument" instead of "short title" or "title". The proper way to define an argument of type "short title" or "title" is to use "short title" or "title" as the whole argument name.
The proposed PR does this and fixes the test error in `LatexParsingTest::test_latexLexing(section command, multi-line optional)`

muzimuzhi commented:
From the package title found on [CTAN](https://ctan.org/pkg/yathesis?lang=en), `yathesis` is
> A LaTeX class for writing a thesis following French rules

Therefore, using a non-English (here the French) argument text maybe be in demand and helpful for specific .cwl file(s). In the long-term and in general, it would be better if texstudio will add support for modifiers like `%short title`, `%title`, and `%envname`. (I should have opened a new issue discussing some more feature requests about `.cwl` files.)

dbitouze commented:
@MeanSquaredError Many thanks for the insight (I'm the author of the `yathesis` class, and the corresponding `.cwl` file)!

>The problem with these definitions is that `%short title` and `%title` are not valid modifiers for the argument type

About `%short title`, I probably misread the TXS' manual, sorry. About `%title`, I don't understand why it isn't valid (from the TXS manual, section 4.14.2 "Command format"):

>`title`, `short title` or ends with `%title`: The spellchecker will operate inside this argument (by default arguments are not spellchecked). Furthermore the argument will be set in bold text (like in section).

muzimuzhi commented:
https://github.com/texstudio-org/texstudio/blob/0245480155e989a0b906fdfabf8021c71c5b9ce3/src/latexpackage.cpp#L629

`%title` is supported. I have also tested in texstudio.

MeanSquaredError commented:
Indeed `%title` is supported (somehow I missed that in the manual), but `%short title` is not.

OK, I will change the patch to add `%short title` and `%envname` instead of modifying yathesis.cwl

MeanSquaredError commented:
OK, added the support for `%short title` and `%envname` in the CWL argument names and documented that in the user manual. Removed the changes to yathesis.cwl

muzimuzhi commented:
> I did not add any specialized tests for `%short title` and `%envname` because there are no definitions in the default CWL files that use the new type modifiers.

The first time I encountered the requirement of supporting `%envname` is in reviewing `float.cwl`, in which
https://github.com/texstudio-org/texstudio/blob/0245480155e989a0b906fdfabf8021c71c5b9ce3/completion/float.cwl#L5
would be better to be record as
```
\newfloat{float type%envname}{placement}{ext}[within]#N
```

For default .cwl files, maybe 
https://github.com/texstudio-org/texstudio/blob/0245480155e989a0b906fdfabf8021c71c5b9ce3/completion/latex-document.cwl#L298-L300
is a candidate to use `%envname`, but I don't see a necessity. Like the previously newly added .cwl syntax (#688), new .cwl syntax mainly required by non-default .cwl files. Therefore, do we need a only-for-test .cwl file, which accumulates all kinds of .cwl syntaxes, or allow use non-default .cwl files for testing?

MeanSquaredError commented:
Right now in the tests we don't have an easy way to turn on/off loaded .cwl files (it might be possible but I have worries about messing up other tests), so we only use the .cwl files that are loaded by default in a new installation of TXS. I think this includes latex-document.cwl and tex.cwl (maybe also latex-dev.cwl or latex-mathsymbols.cwl but I am not sure the latter two ones). So testing environments from float.cwl should be avoided for now (until we have updated the testing framework).

I wouldn't worry too much about the tests - we can manually enable yathesis.cwl and run the tests which (sort of) covers `%short title`. Once someone uses the new type modifiers in latex-document.cwl or tex.cwl we will add the automatic tests.

sunderme commented:
thanks
